### PR TITLE
Refactor code for readability.

### PR DESCRIPTION
**Pull request by Hrothen**
Update code to use Java8 Streams. Broke main function into several
more readable functions. Added a line about Cloud Foundry to the README.
Updated pom.xml to target Java8

### CODE REVIEW

**Comment by colindean** [comment]
lgtm

Do we have any CI set up on this?

File: README.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ The goal of this tutorial is to demonstrate how to get started with an integrati
 Please see the [Introduction](https://github.com/Watson-Explorer/wex-wdc-integration-samples) for an overview of the integration architecture, and the tools and libraries that need to be installed to create Java-based applications in Bluemix.
 
 - An [IBM Bluemix account](https://bluemix.net/)
+- [Cloud Foundry](https://github.com/cloudfoundry/cli/releases)
```
Comment (eoehlsch) [inline README.md]: Cloud foundry is required for all of the WEX-WDC integration examples, and so is included in the list of prereqs discussed at the "Introduction" link above.  If we list Cloud foundry here, we should also list Apache Maven and the IBM JDK.

Comment (Hrothen) [inline README.md]: If there are general requirements for all of the examples, is there a main page where they're listed? Because then we should link back to that.

Comment (eoehlsch) [inline README.md]: That already exists. As I said in my first comment, "and so is included in the list of prereqs discussed at the 'introduction' link above".  I'm talking about the link two lines above.

Comment (Hrothen) [inline README.md]: Huh, I somehow missed that link twice.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 2]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
-		.getJsonObject(0).getJsonObject("credentials");
 
-            try {
-                serviceURL = credentials.getString("url");
-            } catch (Exception e) {}
-            // If we didn't find a URL for the service in VCAP_SERVICES,
-            // use the original.
+
+            serviceURL = credentials.getString("url");
+
```
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Are you sure we can drop the try/catch around credentials.getString() ?  Presumably (I forget exactly) an exception can be thrown if there does not exist a string named "url", but in that case we want to be able to fall back to the ORIGINAL_SERVICE_URL, so we need to catch it.

Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: That will throw an error if "url" does not exist, but do we actually want to "recover" if it's not in the JSON? Something has already gone very wrong.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 3]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
-            conn.setRequestProperty("Transfer-encoding", "chunked");
-	    conn.setFixedLengthStreamingMode(incomingBinaryPOSTData.length);
-
-	    // Uncomment the following to opt out of improving the service with this audio file
-	    //	    conn.setRequestProperty("X-WDC-PL-OPT-OUT", "1");
-            
+
+
+            // set up our http request
+            conn = setupHttpConnection(incomingBinaryPOSTData, serviceURL, credentials);
```
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:55]: incomingBinaryPOSTData can be huge, so this function call made me want to learn (again) whether Java is pass-by-reference or pass-by-value... and I think we are fine.  :)

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 4]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
             // Read the response from the service
-            BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
+            String lines = readResponse(conn);
 
-	    if (DEBUG) {
-		System.out.println("Service response code: " + conn.getResponseCode());
-		System.out.println("Service response message: " + conn.getResponseMessage());
-	    }
+            // Parse the transcript string out of the Speech To Text service's JSON response
+            String transcript = readTranscripts(readJsonObject(new StringReader(lines)));
```
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Would it be more clear to move "readJsonObject(new StringReader(lines))" out of the function call and into the function?

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 5]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+        OutputStream output = conn.getOutputStream();
+        output.write(incomingBinaryPOSTData);
+        output.flush();
+        output.close();
+    }
+
+
+    // Collect the transcripts from all the results in our response body
+    private String readTranscripts(JsonObject jsonResponse) {
+        JsonArray resultsArray = jsonResponse.getJsonArray("results");
```
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: This jsonResponse.getJsonArray("results") used to be inside the try/catch, presumably in the case where there was no array named "results" in the response, so that we can properly blame that on the service.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 6]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+
+        try {
+            return resultsArray.stream().map(jv -> getTranscript((JsonObject) jv)).collect(Collectors.joining());
+        } catch (Exception e) {
+            e.printStackTrace();
+            return failXML(e, "Unexpected response from service caused ");
```
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: I'm not convinced this works... wouldn't this eventually return the failXML text wrapped inside the successXML text?

Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: No, why would it?

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 7]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+        output.close();
+    }
+
+
+    // Collect the transcripts from all the results in our response body
+    private String readTranscripts(JsonObject jsonResponse) {
+        JsonArray resultsArray = jsonResponse.getJsonArray("results");
+
+        try {
+            return resultsArray.stream().map(jv -> getTranscript((JsonObject) jv)).collect(Collectors.joining());
```
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:146]: Hmmm, I'll have to trust you that this is equivalent to the original code...

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 5 above)
Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: You're right, that was an accident on my part.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 2 above)
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Unfortunately, in a cloudy environment, some things can remain working while others fail.  I've seen a case where the service URL was not present in VCAP_SERVICES, but the service itself was still available at the expected URL.
I'd rather try to fall back than give up.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 6 above)
Comment (eoehlsch) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: This method, readTranscripts, is called on line 75.
readTranscripts returns the failXML text to its caller.
Then successXML() is given the output of readTranscripts on line 78.

Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Huh... I guess I accidentally removed a level of nesting from the original try block.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 4 above)
Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Hmm, it actually might be more clear to move it right above the function call; readTranscript conceptually makes more sense as a function that takes a JSONObject.

**Comment by michaelkeeling** [comment]
@colindean - no CI, but that would be awesome and should be 100% doable within Bluemix itself.  We just haven't had the time to fiddle with it.

File: README.md (hunk 1 above)
Comment (michaelkeeling) [inline README.md]: Should use https://github.com/IBM-Watson/wex-appbuilder-samples/blob/master/intro-to-bluemix-for-app-builder.md for introduction to Bluemix for WEX / App Builder.  The old repo is basically abandoned at this point.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 2 above)
Comment (michaelkeeling) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: I think the issue here is that there is a nested try-catch.  VCAP_SERVICES is expected to be populated, the exception will still be caught.  The question is should you try to recover using a known URL.

This is just demo code, but the correct thing to do would be to log the error and also store the "known good but not preferred" URL in a custom environment variable that can be changed without redeploying the app for the recover scenario Eric is worried about.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 3 above)
Comment (michaelkeeling) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:55]: http://stackoverflow.com/questions/40480/is-java-pass-by-reference-or-pass-by-value

TL;DR - everything is pass by value, but you're always passing pointers

File: README.md (hunk 1 above)
Comment (colindean) [inline README.md]: If it's abandoned, it should have a message indicating that it is outdated or it should be deleted.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 8]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
     */
         HttpURLConnection conn = null;
         String serviceURL = "";
-           
+
         try {
-	    if (incomingBinaryPOSTData.length == 0) return XML_DEC + "<results>" + STATUS_OK + "<transcript />" + "</results>";
+            if (incomingBinaryPOSTData.length == 0) {
+                return XML_DEC + "<results>" + STATUS_OK + "<transcript />" + "</results>";
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:43]: I know this is just a reformat, but seeing this raises a question:

Why are we OK with 0 length post?

Whatever the answer is to that could be communicated by extracting that condition to a well-named function, such as `postDataIsSane(incomingBinaryPOSTData)` or something like that.

```
private boolean postDataIsSane(byte[] postdata) { return postData.length == 0; }
```

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 9]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+            }
 
             // 'VCAP_SERVICES' contains all the details of services bound to this application
-	    JsonReader jsonReader = Json.createReader(new StringReader(System.getenv("VCAP_SERVICES")));
-	    JsonObject serviceInfo = jsonReader.readObject();
-	    jsonReader.close();
+            JsonObject credentials = readJsonObject(new StringReader(System.getenv("VCAP_SERVICES")))
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:47]: Should extract VCAP_SERVICES to a constant, especially since `SERVICE_NAME` is a constant below.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 2 above)
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Nested exceptions can be an indicator that another pattern is possible.

``` java
void doSomething() {
  try {
    Result result;
    try {
     result = somethingScary();
    } catch(Exception e) {
      handleException(e);
    }
    if(result == null) { handleFailedResult(); }
    doSomethingWithResult(result);
  } catch(Exception e) {
      handleLargerException(e);
  }
}
```

could be rewritten as

``` java
void doSomething() {
  Result result = getResult();
  ensureSaneResult(result);
  safelyDoSomething(result);
}

void ensureSaneResult(Result result) {
  if(result == null) { handleFailedResult(); }
}

Result getResult() {
  Result result;
  try {
   result = somethingScary();
  } catch(Exception e) {
    handleException(e);
  }
  // could put the ensureSaneResult() sanity check here, too
  return result;
}

void safelyDoSomething(Result result) {
  try {  
    doSomethingWithResult(result);
  } catch(Exception e) {
      handleLargerException(e);
  }
}
```

It's more verbose but its meaning is so much clearer.

In Scala, it might be even cleaner, if the underlying actionable methods played along nicely and used Option and whatnot.

``` scala
def doSomething() = {
  Try(somethingScary).map(doSomethingWithResult).recover(handleException)
}
```

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 10]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
-            output.write(incomingBinaryPOSTData);
-            output.flush();
-	    output.close();
-            
+
+            sendRequest(incomingBinaryPOSTData, conn);
+
             // Read the response from the service
-            BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
+            String lines = readResponse(conn);
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: if `readResponse` is only ever used here, why not make it return a StringReader, or have a method that calls `readResponse` and wraps it in a StringReader?

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 11]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
             // Return the response from the service
-            return XML_DEC + "<results>" + STATUS_OK + "<transcript>" + transcript + "</transcript></results>";
-
-        } catch(Exception e){
+            return successXML(transcript);
+        } catch (Exception e) {
             e.printStackTrace();
-            return XML_DEC + "<results>" + STATUS_ERROR + "<statusInfo>Bluemix " + e.getClass().getName() + ": " + e.getMessage() + "</statusInfo><transcript /></results>";
-        }finally {
+            return failXML(e, "Bluemix ");
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:78]: What's the meaning of Bluemix here? Probably should be a constant, especially if it's used twice: here and on line 89.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 12]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+        // See the Speech to Text service documentation for details.
+        String queryString = "&model=en-US_NarrowbandModel"
+                + "&continuous=true";
+        if (DEBUG) {
+            System.out.println("POSTing [" + incomingBinaryPOSTData.length + "] bytes");
+            System.out.println("Using URL [" + serviceURL + API_PATH + "?" + queryString + "]");
+        }
+
+        // Prepare the HTTP connection to the service
+        HttpURLConnection conn = (HttpURLConnection) new URL(serviceURL + API_PATH + "?" + queryString).openConnection();
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: This could be a function `openConnectionWithQueryString`, mostly to isolate that nasty casting.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 13]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+
+        // Set up our header with all the information we need to make the api request
+        conn.setDoInput(true);
+        conn.setDoOutput(true);
+        conn.setRequestMethod("POST");
+        conn.setRequestProperty("Authorization", "Basic " + Base64.encodeBase64String(auth.getBytes()));
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:118]: I think all of these headers have constants. Use the constants instead.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 14]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+            System.out.println("Using URL [" + serviceURL + API_PATH + "?" + queryString + "]");
+        }
+
+        // Prepare the HTTP connection to the service
+        HttpURLConnection conn = (HttpURLConnection) new URL(serviceURL + API_PATH + "?" + queryString).openConnection();
+
+        // Set up our header with all the information we need to make the api request
+        conn.setDoInput(true);
+        conn.setDoOutput(true);
+        conn.setRequestMethod("POST");
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:117]: Use the POST constant if you can find it.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 7 above)
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:146]: mmmm java 8

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java [hunk 15]
```diff
@@ -43,120 +39,157 @@ public String speechToText(byte[] incomingBinaryPOSTData) {
...
+
+    private String readResponse(HttpURLConnection conn) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
+
+        if (DEBUG) {
+            System.out.println("Service response code: " + conn.getResponseCode());
+            System.out.println("Service response message: " + conn.getResponseMessage());
+        }
+
+        String lines = br.lines().collect(Collectors.joining("\n"));
```
Comment (colindean) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java:163]: TIL `.collect(Collectors.joining())`

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 2 above)
Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: Yeah, the "correct" method is that JsonObject queries should be able to return Optionals instead of throwing. The equivalent Haskell is just:

`result fallbackURL = maybe fallbackURL id . readMaybe`

I think the idea with Java8 Optionals though is that they were supposed to replace Nulls, so we probably won't actually see a rework of libraries to add these methods in Java9 or whatever.

File: README.md (hunk 1 above)
Comment (michaelkeeling) [inline README.md]: We're not going to delete it due to the high number of incoming links.  A note has already been added to the old repo.

File: bluemix/src/main/java/com/ibm/watson/SpeechToText.java (hunk 2 above)
Comment (Hrothen) [inline bluemix/src/main/java/com/ibm/watson/SpeechToText.java]: After all that it turns out that there's an overload of getString() that takes a fallback value.
